internal/provider/serverless_branch_data_source.go

```diff
@@ -0,0 +1,253 @@
+	CreatedBy         types.String `tfsdk:"created_by"`
+	CreateTime        types.String `tfsdk:"create_time"`
+	UpdateTime        types.String `tfsdk:"update_time"`
+	ParentDisplayName types.String `tfsdk:"parent_display_name"`
+	ParentTimestamp   types.String `tfsdk:"parent_timestamp"`
+	Annotations       types.Map    `tfsdk:"annotations"`
+}
+
+var _ datasource.DataSource = &serverlessBranchDataSource{}
+
+type serverlessBranchDataSource struct {
+	provider *tidbcloudProvider
+}
+
+func NewServerlessBranchDataSource() datasource.DataSource {
+	return &serverlessBranchDataSource{}
+}
+
+func (d *serverlessBranchDataSource) Metadata(_ context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_serverless_branch"
+}
+
+func (d *serverlessBranchDataSource) Configure(_ context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+	var ok bool
+	if d.provider, ok = req.ProviderData.(*tidbcloudProvider); !ok {
+		resp.Diagnostics.AddError("Internal provider error",
+			fmt.Sprintf("Error in Configure: expected %T but got %T", tidbcloudProvider{}, req.ProviderData))
+	}
+}
+
+func (d *serverlessBranchDataSource) Schema(_ context.Context, _ datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		MarkdownDescription: "serverless branch data source",
+		Attributes: map[string]schema.Attribute{
+			"cluster_id": schema.StringAttribute{
+				MarkdownDescription: "The ID of the cluster.",
+				Required:            true,
+			},
+			"branch_id": schema.StringAttribute{
+				MarkdownDescription: "The ID of the branch.",
+				Required:            true,
+			},
+			"display_name": schema.StringAttribute{
+				MarkdownDescription: "The display name of the cluster.",
+				Computed:            true,
+			},
+			"parent_id": schema.StringAttribute{
+				MarkdownDescription: "The parent ID of the branch.",
+				Computed:            true,
+			},
+			"endpoints": schema.SingleNestedAttribute{
+				MarkdownDescription: "The endpoints for connecting to the cluster.",
+				Computed:            true,
+				Attributes: map[string]schema.Attribute{
+					"public": schema.SingleNestedAttribute{
+						MarkdownDescription: "The public endpoint for connecting to the cluster.",
+						Optional:            true,
+						Computed:            true,
+						Attributes: map[string]schema.Attribute{
+							"host": schema.StringAttribute{
+								MarkdownDescription: "The host of the public endpoint.",
+								Computed:            true,
+							},
+							"port": schema.Int32Attribute{
+								MarkdownDescription: "The port of the public endpoint.",
+								Computed:            true,
+							},
+							"disabled": schema.BoolAttribute{
+								MarkdownDescription: "Whether the public endpoint is disabled.",
+								Optional:            true,
+							},
+						},
+					},
```

Comment:
no optional I think. And use branch instead cluster in description. Also private below.

Reply:
also in branches